### ISSUE TITLE
Summary of book tests

### ISSUE THREAD

**Issue by vdbergh** [official-stockfish/Stockfish#3323]
The book tests are finished so I am summarizing the results here. From a technical point of view the only column that is important is "Normalized Elo" since it determines the number of games needed to detect the given strength difference (the relation is inversely quadratic, for SPRT the approximate formula is `640000/(n.e.)^2`)

For a similar series of test (reported on Jan 3, 2020) see https://github.com/glinscott/fishtest/pull/472#issuecomment-570553491 . The issue in loc. cit. also gives more information about the various books. The books themselves can be found in the [Fishtest book repository](https://github.com/official-stockfish/books).

| Name | Draw ratio | Bias(Elo) | Elo | Normalized Elo | Test |
|-|-|-|-|-|-|
|Drawkiller_balanced_big.epd (LTC)|0.63|70| 70.6 [69.0, 72.1]|128.6 [125.6, 131.6]|[Test](https://tests.stockfishchess.org/tests/stats/600c5ed8735dd7f0f0352bad)|
|bjbraams_chessdb_198350_lines.epd (LTC)| 0.65|     122 | 57.3 [56.0, 58.6] | 123.1 [120.0, 126.2] |[Test](https://tests.stockfishchess.org/tests/stats/610438b8afad2da4f4ae3b8a)|
|hybrid_book_beta.pgn (LTC)|0.70|66|59.7 [58.3, 61.1]|121.0 [118.2, 123.9]|[Test](https://tests.stockfishchess.org/tests/stats/600d4342735dd7f0f0352c33)|
|UHO_XXL_+0.80_+1.09.epd (LTC)|0.55|165|57.4 [56.0, 58.8]|119.0 [115.9, 122.1]|[Test](https://tests.stockfishchess.org/tests/view/611923594977aa1525c9c77f)|
|UHO_XXL_+0.90_+1.19.epd (LTC)|0.49|189|57.6 [56.2, 58.9]|118.0 [114.9, 121.1]|[Test](https://tests.stockfishchess.org/tests/view/611923644977aa1525c9c781)|
|8mvs_big_+80_+109.epd (LTC)|0.56|156|56.6 [55.2, 58.0] | 116.0 [112.9, 119.1]|[Test](https://tests.stockfishchess.org/tests/view/60112110735dd7f0f0352df9)|
|2moves_v1.pgn (LTC)|0.76 |77 | 47.0 [45.8, 48.2]| 110.5 [107.7, 113.3] |[Test](https://tests.stockfishchess.org/tests/stats/600c2f74735dd7f0f0352b97)|
|UHO_XXL_+1.00_+1.29.epd (LTC)|0.43|219|52.9 [51.5, 54.3]|108.7 [105.6, 111.8]|[Test](https://tests.stockfishchess.org/tests/view/611923724977aa1525c9c783)|
|closedpos.epd (LTC)|0.79|54|42.0 [40.8, 43.2]|101.7 [99.0, 104.5]|[Test](https://tests.stockfishchess.org/tests/stats/607686c28141753378960681)|
|noob_3moves.epd (LTC)|0.83| 23| 37.6 [36.5, 38.7]| 96.9 [94.3, 99.5]|[Test](https://tests.stockfishchess.org/tests/stats/600c1767735dd7f0f0352b8b)|
|8moves_v3.pgn (LTC)|	0.83|58| 32.4 [31.4, 33.4]| 	87.6 [84.9, 90.2]|[Test](https://tests.stockfishchess.org/tests/stats/6000a2616019e097de3ef77d)|
|endgames.epd (LTC)|0.63|181|24.6 [23.6, 25.6]|66.5 [63.8, 69.3]|[Test](https://tests.stockfishchess.org/tests/stats/6075b6818141753378960617)|

Some STC tests to check scaling.

| Name | Draw ratio | Bias(Elo) | Elo | Normalized Elo | Test |
|-|-|-|-|-|-|
|Drawkiller_balanced_big.epd (STC)|0.47|62| 86.8 [84.8, 88.8] | 130.9 [127.8, 134.0] |[Test](https://tests.stockfishchess.org/tests/stats/60805c6c111fbbda54e11771)|
|2moves_v1.pgn (STC)|0.63| 77  | 64.3 [62.8, 65.9]| 118.6 [115.6, 121.5]|[Test](https://tests.stockfishchess.org/tests/stats/60850dd195e7f1852abd27fa)|
|bjbraams_chessdb_198350_lines.epd (STC)| 0.57|     121 |63.8 [62.2, 65.3]   | 117.7 [114.7, 120.83 |[Test](https://tests.stockfishchess.org/tests/stats/60fdc57ad8a6b65b2f3a79d8)|
|UHO_XXL_+0.80_+1.09.epd (STC)|0.50|154|63.0 [61.4, 64.6]|114.8 [111.7, 117.9]|[Test](https://tests.stockfishchess.org/tests/view/611922fb4977aa1525c9c779)|
|noob_3moves.epd (STC)|0.70| 32| 59.7 [58.2, 61.2] | 114.2 [111.7, 117.1]|[Test](https://tests.stockfishchess.org/tests/stats/60805bb9111fbbda54e1176f)|
|UHO_XXL_+0.90_+1.19.epd (STC)|0.46|171|61.8 [60.3, 63.4]|112.3 [109.2, 115.4]|[Test](https://tests.stockfishchess.org/tests/view/611923194977aa1525c9c77b)|
|UHO_XXL_+1.00_+1.29.epd (STC)|0.43|193|60.0 [58.5, 61.6]|110.3 [107.2, 113.4]|[Test](https://tests.stockfishchess.org/tests/view/6119232d4977aa1525c9c77d)|
|8moves_v3.pgn (STC)|0.71| 61  | 49.0 [47.6, 50.4]| 98.8 [95.9, 101.7] |[Test](https://tests.stockfishchess.org/tests/stats/60850e1295e7f1852abd27fc)|

EDIT: Added 8mvs_big_+80_+109.epd. This book with 25857 positions is the biggest in the series Unbalanced_Human_Openings_V2.0, created by Stefan Pohl. See https://www.sp-cc.de/unbalanced-human-openings.htm.

EDIT2: Added data for the special purpose books "endgames.epd" and "closedpos.epd".

EDIT3: I started adding STC data to check scaling.

EDIT4: Some more STC data. I did the tests a long time ago but never got around adding the results.

EDIT5: STC data for bjbraams_chessdb_198350_lines.epd added.

EDIT6: LTC data for bjbraams_chessdb_198350_lines.epd added. Strong performance.

EDIT7: Data for the new UHO series by Stefan Pohl.

**Comment by Alayan-stk-2** [comment]
It would be interesting to see a comparison test with Pohl's UHO : https://www.sp-cc.de/unbalanced-human-openings.htm

They probably lack some variety coverage, but they are a good example of openings selected to be close to the draw/win limit. Such openings are worse at differentiating engines of very different strength, or weak engines, but in tournament conditions with big hardware they are very good compared to balanced openings.

I'm not sure where fishtest LTC currently sits, but the draw rates are so high than making the possibility of a "weak side win" really tiny might not matter.

**Comment by vdbergh** [comment]
Yes that would be interesting.

Before a book can be tested it has to be uploaded to the book repository. I guess I can just make a pull request...

**Comment by zz4032** [comment]
Isn't there a correlation between `Normalized Elo` and `Draw ratio`?
It looks like a book is generally more appropriate for engine testing, if
- `Draw ratio` is lower / closer to 0.50?,
- `Bias(Elo)` is lower (more balanced)?,
- it includes more realistic openings (cannot really be expressed in numbers)?

**Comment by ddobbelaere** [comment]
> * it includes more realistic openings (cannot really be expressed in numbers)?

You can define it as how similar it is to opening distribution from e.g. reference database of human games. E.g. https://en.m.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence

**Comment by vdbergh** [comment]
Here is a similar series of tests (originally initiated by Vondele) for a much larger Elo difference (which includes in particular the 100 Elo jump due to the introduction of NNUE). One can observe that Drawkiller still wins...  This book is definitely interesting.

The column Bias should be taken with a grain of salt since the model to compute it assumes small Elo differences, and moreover there is not enough information to compute a confidence interval. 

| Name | Draw ratio | Bias(Elo) | Elo | Normalized Elo | Test |
|-|-|-|-|-|-|
|Drawkiller_balanced_big.epd|0.26|41|271.3 [268.7, 273.9]|  426.0 [420.3, 431.9]|[Test](https://tests.stockfishchess.org/tests/view/6011232e735dd7f0f0352dfb)|
|noob_3moves.epd (STC)|0.28|7|270.9 [268.27, 273.5] |       416.3 [410.6, 422.0]|[Test](https://tests.stockfishchess.org/tests/view/600fbb91735dd7f0f0352d57)|
|noob_3moves.epd|0.39|35|227.7 [225.6, 229.9]|              385.4 [380.8, 390.0]|[Test](https://tests.stockfishchess.org/tests/view/600fbb9c735dd7f0f0352d59)|
|8moves_v3.pgn|0.47|64|182.9 [181.0, 184.7]|                328.5 [324.6, 332.4]|[Test](https://tests.stockfishchess.org/tests/view/600fbbb7735dd7f0f0352d5d)|

**Comment by vdbergh** [comment]
> Isn't there a correlation between `Normalized Elo` and `Draw ratio`?
> It looks like a book is generally more appropriate for engine testing, if
> 
> * `Draw ratio` is lower / closer to 0.50?,
> * `Bias(Elo)` is lower (more balanced)?,
> * it includes more realistic openings (cannot really be expressed in numbers)?

Before NNUE I think there was no such direct correlation. IIRC noob_3moves.epd and 2moves_v1.pgn had similar sensitivity, despite having radically different properties.

Since the introduction of NNUE the draw ratio has gone through the roof and it seems the resulting Elo compression can no longer be compensated for by the reduced noise associated with a higher draw ratio.

**Comment by vdbergh** [comment]
Added the results for the "Unbalanced Human Openings" by Stefan Pohl to the OP.

**Comment by zz4032** [comment]
If you're interested in testing books for lower draw rates but with more positions (and without hand-selection of specific types of positions): I have some of my own 3-move-books that I have generated with SF NNUE and grouped by evaluation:
[3moves_cp.zip](https://github.com/official-stockfish/Stockfish/files/5888078/3moves_cp.zip)
`3moves_cp50-79_152057pos` or `3moves_cp80-112_101530pos` could be nice candidates for higher `Normalized Elo`, while not too unbalanced.

**Comment by vdbergh** [comment]
@zz4032 I didn't really want to test books perse (it takes a lot of resources). I just wanted to verify the suggestion by many people that the draw ratio may have become so high that an ulta balanced book like noob_3moves.epd is no longer the best choice for Fishtest. It seems that the tests confirm this.

**Comment by vondele** [comment]
@vdbergh thanks for starting this investigation. I added the additional tests mostly because they reflect one year of progress, coinciding with 1yr of my role as maintainer. Progress has been unimaginable..

Back to your tests... first, I think we should really push for pentanomial statics and normalized Elo as basic tools to use for high level computer chess. I have the feeling these are the right tools to use at this level. I would be in favor of making that more explicitly used in fishtest.

Concerning books, let me say I do like our noob_3moves.epd, so I'll be a bit biased. 

Looking at normalized Elo, books seem a little closer to each other than in Elo, which I think is a good thing. Yet, one could argue that there is a 2x gain possible by moving to a different book (in number of games needed to decide the typical SPRT). Having a large (normalized) Elo is clearly an important property of a book, but is not the only one. Those need considering as well:

* book size: Ideally each game played in fishtest starts from a different opening, to reduce correlations between games played. Since we quite commonly play 100k - 300k games, books should be 50k - 150k positions. 
* book content: the engine will naturally be optimized for the type of positions that are in the book. If our book is too specific, we will become pretty specific in what we learn, and become relatively speaking worse in other context.

While book size is quantifiable easily, book content is not. I would argue that having a book that is mostly reasonable positions close to startpos is a good approach to train the engine for generically strong play. I know there are arguments for other kind of books. 

Another question is, what would we like to achieve with a book change, and could we achieve this by other means as well (obviously with other side effects). In particular, SPRT bounds could be adjust to make patches easier to pass (and moved to normalized Elo, so we can keep them constant for longer, and more properly adjusted to the kind of test running), and maybe we could consider changing TC (Elo differences are larger at lower TC as well). The later point is maybe controversial, but we moved to 60s to have high quality of play.... our 10s now is miles better play than we had at 60s when we introduced it.

**Comment by miguel-l** [comment]
Slightly off topic, but might give some ideas..
My idea for nice book content is something like a mix of the following:

- chess 960 - this is probably already difficult, since afaik chess960 is not supported by fishtest, but I feel like this would help very much as middlegames that arise from it will varied.
- Modified starting positions - once you get the above this is pretty easy to generate by modifying fens. Some examples of what I'm thinking: Take all 960 fens and remove a rook from one side and a bishop from the other, Or remove both queens, Or force OCB from the start, Or remove queenside castling rights for white and kingside for black, etc, etc. I think this would cover imbalanced and positions close to endgames. Programmatic modification of fens might end up with 'impossible' positions, but it might not matter much.
- Positions from other books - Take 2 or 3moves or some other existing book and take positions from them. This should boost the distribution so that the book still contains many positions that are close to standard chess.

**Comment by NKONSTANTAKIS** [comment]
One problem for books with low resolution, or to be more precise with resolution which drops significantly as TC increases, is that the search optimizations will be too specialized for fishtest conditions and losing value.

High resolution positions pose a difficult puzzle, so its much more beneficial to tune for them. In other words you would prefer to maximize performance for needing less time to not make crucial mistakes, than to do less mistakes on average when we use low time.

So lowering LTC is opposing, and high resolution aiding.

Being generic with book is of course also an asset, helping the optimization to be more generically applicable.
So, even though we dont tune eval, concerns for overly specialized max resolution book, like drawkiller are understood. If even if we don't want to research the generic applicability of tuning for drawkiller and go safe, why not to aim for both assets of genericality and resolution, just by usage of normal positions?

Having the whole book at the same length reduces genericality. Why not to construct a variable length book by taking the highest resolution part of all normal books?

Also, I am very excited as the importance of the concept of the scaling of resolution was made clear to me just now. I will attempt to explain further with example.

So we have 2 books, one with slightly lower resolution but more uniform across TC and 2nd with better but with LTC drop noticeably overlapping elo compression. Using the first will mean:

1. That the STC result will be more correlated with LTC result, thus increased confidence.
2. That the general optimization will scale better with time.
3. That the general optimization will be more uniform.

So I think its valuable and cheap to measure the scaling of resolution for books.

**Comment by NKONSTANTAKIS** [comment]
Would it make sense to use more than 1 book and adjust normalized bounds in relation to their resolution too?
This way the target test length can be similar, cross book performance comparable, and target midpoint elo different (lower for low res, higher for high).

Because I'm not that sure that just with normalized elo plus adjusting confidence accordingly is the same. That would mean that resolution is unimportant to selectability and economy, and just position coverage essential as @vondele hinted. This idea is interesting.

In other words, even with 98% drawrate (as long as there are no openings which produce 100% and waste resources) progress rate and test length could be kept stable just by reducing the win-loss demand. This somehow is hard to believe, but if to an extent true then lowering LTC fully makes sense.

**Comment by Alayan-stk-2** [comment]
> Another question is, what would we like to achieve with a book change, and could we achieve this by other means as well (obviously with other side effects). In particular, SPRT bounds could be adjust to make patches easier to pass (and moved to normalized Elo, so we can keep them constant for longer, and more properly adjusted to the kind of test running), and maybe we could consider changing TC (Elo differences are larger at lower TC as well). The later point is maybe controversial, but we moved to 60s to have high quality of play.... our 10s now is miles better play than we had at 60s when we introduced it.

I think the idea of shortening LTC because quality of play has increased a lot is forgetting that testing at a longer TC isn't primarily about having test games above some absolute chess strength.

While the observation that what works at short TC also often helps at longer TC has been key to generalize mass statistical testing that made a lot of engine progress possible ; there is still a distortion between what is optimal for different TC/hardware budgets. The scaling of the search tune in https://github.com/official-stockfish/Stockfish/pull/2260 was a stark reminder of this fact.

The use cases where Stockfish's strength matter most are those with deep searches - analysis, engine tournaments, etc. Shortening the TC would increase the distortion and while it would still be gaining strength at all TCs, the testing process would statistically optimize less for the long searches that matter the most.

**Comment by Vizvezdenec** [comment]
Yes, there is a problem with having shorter LTC.  
Problem is that search in a lot of cases is non-linear, I can name you like 5 different ideas that were gods of STC but at LTC they were utter trash.
Risk matrix, skipping fail high resolve, lmr history, lowering counter move history pruning threshold - this is what I get by 5 seconds of digging in my memory :)
If anything I would prefer maybe ever higher LTC but with much lower draw rate. I don't know how bias increase affect anything, tbh, so I'm not sure that let's say hybrid book beta will be good for it.

**Comment by vdbergh** [comment]
I added some more STC data. 

If ever the framework becomes empty again then it might be a good idea to do  some tests with a more recent pair of stockfishes. Of course the normalized Elo values will be different but one may hope that the relative ratios between the different books remain similar.

**Comment by vondele** [comment]
@vdbergh I ran a comparable test on a new book very recently. Could you add it https://tests.stockfishchess.org/tests/view/60fdc57ad8a6b65b2f3a79d8

**Comment by vdbergh** [comment]
@vondele 

Done. The book is on par with 2moves_v1.pgn and noob_3moves.epd.  It's a very small sample size but it seems plausible that at STC any "reasonable" book (not too many moves, good coverage) essentially has the same resolution (independently of draw ratio or bias, as long as they are not extreme). The draw killer book cheats by reducing coverage (the issue of castling vs not castling cannot occur in a game started from a draw killer opening).

**Comment by vdbergh** [comment]
It could of course be that Elo compression does exist for the very high draw ratios in currently Fishtest (even when normalized Elo is used) which we are not seeing in these tests. This is not so easy to test since one needs to use engines which are fairly close together in strength, but then one needs a very large number of games to get sufficiently precise results.

**Comment by vondele** [comment]
I'll do an LTC test for that new book as well. I'll start it once the current RT is done.

**Comment by vdbergh** [comment]
I added the LTC data for bjbraams_chessdb_198350_lines.epd to the table. This seems like an interesting book!

**Comment by SFisGOD** [comment]
Let's make bjbraams_chessdb_198350_lines.epd the new default book! :)

**Comment by vdbergh** [comment]
To make theoretical prediction about the merits of biased books one needs a draw model and the validity of the conclusions then depends on the validity of the model. This being said, assuming the BayesElo model one obtains the following picture.

![img](https://hardy.uhasselt.be/Fishtest/sensitivity.png)

Here the "draw ratio" is the one which would occur with a perfectly balanced book (the noob_3moves book is a good approximation).

**Comment by vdbergh** [comment]
To make theoretical prediction about the merits of biased books one needs a draw model and the validity of the conclusions then depends on the validity of the model. This being said, assuming the BayesElo model one obtains the following picture.

https://hardy.uhasselt.be/Fishtest/sensitivity.png

Here the "draw ratio" is the one which would occur with a perfectly balanced book (the noob_3moves book is a good approximation).

EDIT: I tried to include the picture directly in this comment but for some reason this does not work (even though I have done it many times in the past). Anyway below is the command. Perhaps it will start working magically in the future

![Broken picture](https://hardy.uhasselt.be/Fishtest/sensitivity.png)

**Comment by vdbergh** [comment]
Here is another interesting picture. It says that the optimal bias occurs when the draw ratio is 50% (still assuming the BayesElo model). This is rather appealing from the point of maximum entropy (but I see no a priori reason why the entropy argument would be valid).

https://hardy.uhasselt.be/Fishtest/draw_ratio.png

Win and loss ratios are for the player who has the positive bias. Since the openings are repeated with reversed colors, the total win and loss ratios will be close to 25% at the "optimal bias" (assuming the engines are equal strength).

**Comment by NKONSTANTAKIS** [comment]
By my observations in practice, I would say that 50% is not optimal, but 60%-80% is. The optimal number however for sure will be different for different engine pairs and different books.

@vdbergh Regarding maximum information extraction I have the following proposition, which is unclear to me if its throughout sufficient or partially viable: **50% could be optimal in case that average game length of decisive and drawn games is the same.**

But of course its expected that drawn games last longer on average than decisive ones, so they put more tests to the engines, albeit easier ones, hence more information.
An extreme example of the opposite case are positions which require just a single good move to be found, and this move is found 50% of the time. So this 50% derives from a very good test, but just a single one.
Early forced draws by repetition/perpetual etc. fall in this category.

So, to sum up, sustained complexity (unclear outcome) of positions is also important, along with draw rate. We would prefer positions that resolve into draw/win at long games, than positions which resolve fast.

**Comment by NKONSTANTAKIS** [comment]
Regarding bjbraams_chessdb_198350_lines.epd , I find most impressive to have the least decline of elo spread from STC to LTC, indicating sustained uncertainty. (As opposed to increase of determinism with TC).  For sure this asset will be even more important at higher drawrates, stronger engines, and closer in strength engine pairs, than the ones tested here.

I too think it will be beneficial to start using it, in order to also observe its behavior at target conditions.

**Comment by vdbergh** [comment]
I did the computation for the Davidson model 

https://hardy.uhasselt.be/Fishtest/Paired%20Comparisons%20with%20Ties_%20Modeling%20Game%20Outcomes%20in%20Chess.pdf

and the results are very similar to the ones for the BayesElo model (of course these are just _models_, they were mainly selected for their mathematical elegance and it is not clear how well they agree with reality, especially under extreme conditions).

https://hardy.uhasselt.be/Fishtest/draw_ratio_davidson.png

The optimal draw ratio is still around 50%. The gain obtainable from using biased books is still huge.

**Comment by vdbergh** [comment]
Interesting. It seems that any model which has the property that the win ratio is given by `f(elo+bias-draw_elo)` where is `f` is some (sufficiently nice) ascending function satisfying `f(elo)=1-f(-elo)` has the property that for large `draw_elo`, the optimal bias is given by `bias=draw_elo` which corresponds to a draw ratio of `50%`.

This assumption holds on the nose for Glen-Davis and BayesElo. 

It also holds asymptotically for Davidson, as one may check that BayesElo (officially Rao-Kupper) and Davidson are equivalent for large biases.

**Comment by vdbergh** [comment]
For the BayesElo model I found a nice approximative formula for the conversion `draw ratio->normalized Elo`. Namely
```
n.e. ~ sqrt(d*(1-d))
```
(`d=draw ratio`, the symbol `~` indicates proportionality).

**Example** Assume we use a book which reduces the draw ratio from `94.3%` to `50%`. Then according to the BayesElo model, this would magnify normalized Elo (for engines which are close together in strength) with a factor
```
sqrt(0.5*(1-0.5))/sqrt(0.943*(1-0.943))=2.16
```
 which corresponds to what is shown on the graphs.

For larger draw ratios between equal engines the formula also works for the Davidson model (say for `d>85%`).

**Disclaimer** I have no idea how well the BayesElo/Davidson models correspond to reality, especially for extreme draw ratios such as `94.3%`. So for now this should be regarded as a pleasant mathematical exercise.

**Comment by vdbergh** [comment]
In made a short note on the derivation of the formula 
```
n.e. ~ sqrt(d(1-d))
```
https://hardy.uhasselt.be/Fishtest/draw_ratio_to_normalized_elo.pdf

**Comment by bjbraams** [comment]
For my own understanding of the value of bias and of optimal bias I find it attractive to think of a simple binary test between two engines to decide which one is stronger without trying to decide how much stronger. With use of opening book each two-game minimatch constitutes an experiment or measurement of which engine is stronger. The experiment succeeds if the minimatch is decisive and it fails if we get a drawn minimatch. So other things (time control) being equal we want to maximize the fraction of decisive minimatches.
Now we need a model, but here I offer only a handwaving argument. If the engines are very strong and are close in strength then each selected opening allows in practice only two outcomes: win by the advantaged side or draw. Assign probabilities, optimize, and it is seen that one wants a draw ration of 0.5, leading to a fraction of decisive minimatches also 0.5 (and win-win by same colour and draw-draw both at 0.25).
The result that bias should be selected to obtain draw ratio close to 0.5 is not universal. If the engines are far apart in strength and/or the engines are a bit erratic then an optimal bias would be close to zero, giving the stronger engine a chance to win with either colour. I think that models will exhibit a bifurcation in the result; either high bias and a draw ratio close to 0.5 or low bias.
Of course there may be other criteria that compete with the bias and the draw ratio, for example familiarity from human play or (in the TCEC context) expected viewer excitement.

**Comment by vdbergh** [comment]
The entropy argument is not valid. There is no reason to assume that a binary outcome with equal probabilities yields maximal information. What if the outcome is the result of tossing a fair coin? This not a ridiculous comparison since lowering the draw ratio actually **increases noise**!

I find a derivation starting from a reasonable Elo model a lot more convincing. But since the Elo models are hypothetical, only measuring provides the definite answer.

**Comment by bjbraams** [comment]
My reasoning that is focused on the minimatch draw rate and on the binary test of which engine is stronger seems to give answers close to what @vdbergh obtains from a more detailed model. Take the example presented earlier: "Assume we use a book which reduces the draw ratio from 94.3% to 50%." Now, draw ratio on its own does not determine minimatch draw ratio, but under suitable assumptions for draw ratio d I get minimatch draw ratio d^2+(1-d)^2; for draw ratio 94.3% minimatch draw ratio 89.25% or decisive fraction 10.75%. Still under suitable assumption, for draw ratio 50% I get minimatch draw ratio 50% and decisive fraction also 50%. Taking the ratio 0.50/0.1075 I obtain 4.65 as the achievable gain, and @vdbergh quoted 4.66 for that factor. Close enough for TCEC work, I say.

**Comment by vondele** [comment]
Meanwhile Pohl published his new set of books: https://www.sp-cc.de/uho_xxl_project.htm

and I have started some tests (3 books, STC and LTC) for them here:

https://tests.stockfishchess.org/tests/view/611922fb4977aa1525c9c779
https://tests.stockfishchess.org/tests/view/611923194977aa1525c9c77b
https://tests.stockfishchess.org/tests/view/6119232d4977aa1525c9c77d
https://tests.stockfishchess.org/tests/view/611923594977aa1525c9c77f
https://tests.stockfishchess.org/tests/view/611923644977aa1525c9c781
https://tests.stockfishchess.org/tests/view/611923724977aa1525c9c783

**Comment by vondele** [comment]
I'm wondering about the draw ratio discussion. In some sense draw ratio is a 'trinomial' concept, and should be flawed. What about a book where all openings are either guaranteed draws or guaranteed wins. The draw ratio can be 50% yet the Elo difference is 0? Probably this needs to be rephrased with the proper pentanomial concepts?

**Comment by bjbraams** [comment]
Yes, it is correct to focus on the pentanomial concepts. I don't know if you (@vondele) have any particular place in mind where this should be addressed. I try to emphasize on any suitable occasion that the minimatch draw ratio, and not the simple games draw ratio, is the proper object of concern. Lately I have been adding the word "pentanomial" when I speak of minimatches in the fishtesting context. The key issue for a mathematical treatment is that games are strongly pairwise (anti-)correlated whereas minimatches can more reasonably be viewed as independent measurements.
I think that we all understand this, as does anyone associated with the quite sophisticated fishtest SPRT work. Certainly I see no disagreement between @vdbergh and myself; as seen in an earlier contribution here we are finding the same achievable gain through a sophisticated (@vdbergh) and a simple model, pointing to the correctness of our various assumptions. (This, at least, in the presently relevant fishtesting context where all the time two strong and very close engines are being compared.)

**Comment by vondele** [comment]
The UHO book tests finished, and look pretty good to me. Re-evaluating a few thousand positions with SF shows these graphs:

![image](https://user-images.githubusercontent.com/4202567/129672863-0a0f207c-c089-4fe6-9376-6241837d65c0.png)

with a distribution around 138cp, quite independent on the depth used to rescore. From the win_rate_model, 138 is close to an expected 50% win probability. (see also in discord https://discord.com/channels/435943710472011776/813919248455827515/876711431922384966 )

**Comment by vondele** [comment]
Additional test of the UHO_XXL_+0.90_+1.19.epd book, using a net update test:

noob_3moves.epd
https://tests.stockfishchess.org/tests/view/611b58584977aa1525c9c8eb
https://tests.stockfishchess.org/tests/view/611b67e94977aa1525c9c8f7


UHO_XXL_+0.90_+1.19.epd
https://tests.stockfishchess.org/tests/view/611bf0334977aa1525c9c962
https://tests.stockfishchess.org/tests/view/611c065b4977aa1525c9c96a

quite interesting result

**Comment by vdbergh** [comment]
> Meanwhile Pohl published his new set of books: https://www.sp-cc.de/uho_xxl_project.htm
> 
> and I have started some tests (3 books, STC and LTC) for them here:
> 
> https://tests.stockfishchess.org/tests/view/611922fb4977aa1525c9c779
> https://tests.stockfishchess.org/tests/view/611923194977aa1525c9c77b
> https://tests.stockfishchess.org/tests/view/6119232d4977aa1525c9c77d
> https://tests.stockfishchess.org/tests/view/611923594977aa1525c9c77f
> https://tests.stockfishchess.org/tests/view/611923644977aa1525c9c781
> https://tests.stockfishchess.org/tests/view/611923724977aa1525c9c783

I am on vacation till tomorrow. I will add the results to the table when I get back.

@vondele

We should consider doing similar tests with commits that are closer together in strength so that the draw ratio of the noob book is higher (for really low differences in strength, we may have to make the test longer than 60000 games to get significant results).

The reason is that it is possible that the savings from lowering the draw ratio become more pronounced  with higher draw ratio. This is what standard Elo models suggest (cfr my posts above and also
https://hardy.uhasselt.be/Fishtest/draw_ratio_to_normalized_elo.pdf )
but we have currently no systematic experimental evidence to confirm this.

**Comment by vondele** [comment]
yes, let's see them in a table once you're back.

Another test of a normal test redone with the UHO book:
noob:
https://tests.stockfishchess.org/tests/view/611c6c7d4977aa1525c9c996
UHO:
https://tests.stockfishchess.org/tests/view/611c9b214977aa1525c9c9cb

**Comment by noobpwnftw** [comment]
It looks like the main difference is that the UHO books removed very balanced lines and always gives white side an advantage, with my book I did not filter white lines which started with a small disadvantage or simply neutral. If coverage is not of any concern(which is a mystery anyways, I have yet to see any examples showing this could've caused a bias) then I think it is fine to use the new book.

**Comment by vdbergh** [comment]
If the positive bias is always for white then this should be remembered since it is conceivable that there are colour dependent patches that would pass due to the dependence between bias and colour.

**Comment by vdbergh** [comment]
Data added for the new UHO series by Stefan Pohl. At STC there no difference in performance with noob_3moves. At LTC there is a gain in performance.

**Comment by vdbergh** [comment]
I started a new series of tests with engines that are ~ 10 (logistic) Elo apart.

https://tests.stockfishchess.org/tests/view/611fdc804977aa1525c9cba6  (noob_3moves.epd)
https://tests.stockfishchess.org/tests/view/611fdd2b4977aa1525c9cba9 (UHO_XXL_+0.90_+1.19.epd)
https://tests.stockfishchess.org/tests/view/612095c44977aa1525c9cc07 (bjbraams_chessdb_198350_lines.epd)

The tests are far from finished, but at this point it is actually the noob book that is ahead in normalized Elo (which is what matters for fishtest)!! Of course it is to early to discuss this.

**Comment by NKONSTANTAKIS** [comment]
@vdbergh The 2 versions that are 10 elo apart are not neutral elo, but elo based on performance on noob_3moves. So this book's performance is boosted, to an unknown degree. 
In order to have a fully neutral comparison I recommend to use the exact same version against itself and create the desired elo difference via time odds.

**Comment by vdbergh** [comment]
> @vdbergh The 2 versions that are 10 elo apart are not neutral elo, but elo based on performance on noob_3moves. So this book's performance is boosted, to an unknown degree.
> In order to have a fully neutral comparison I recommend to use the exact same version against itself and create the desired elo difference via time odds.

I agree that selection bias for the noob book is a genuine risk. On the other hand if it exists then it was not enough to make the noob book dominate the test in the OP.

Time odds is an interesting possibility. I am not sure how to do it though. I don't particularly like nodeistime.

**Comment by NKONSTANTAKIS** [comment]
I worry in particular due to their vastly different nature: one has quiet positions with 94-95% draw-rate in master self-play, and the other extreme borderline with 48-50%. So optimization could potentially be very different, and all has been done for the first case.

I am also not aware how time odds are done properly at framework, feedback required. @vondele @locutus2 @xoto10 @SFisGOD

**Comment by SFisGOD** [comment]
Using Slow Mover like this one https://tests.stockfishchess.org/tests/view/611636e5055f06d4c60d08e0

**Comment by NKONSTANTAKIS** [comment]
Any clue about what value would give 10-15 elo weakening?

**Comment by xoto10** [comment]
I did test some different values (earlier than this test) but not sure I can find them now. I don't know the shape of the graph so we could just assume linear, do a test and then refine. I suspect the highest values make little difference, and then when you get down to 25% of time like this smaller changes are quite noticeable ... Linear would suggest ~ 85%, so maybe it's more like 75%?
I requested a couple of tests with 70 and 80 ...

**Comment by xoto10** [comment]
It may be worth noting that the Slow Mover option just makes stockfish take a smaller amount of time than it normally would. The same total time given by the time control is available, so this makes it move quicker in the early part of the game, conserving time for later. Of course, it still does this later on, conserving time for ... even later on ... but at small decrements, the extra time conserved after a few moves may balance out the decrement imposed by the option.

It could perhaps be viewed as a "Bad Time Management" option rather than actual time odds.

The advantage of using it is that is already built in, and allows easy testing of any commit against itself.

**Comment by NKONSTANTAKIS** [comment]
Hm, I thought this kind of bias would be relatively harmless, but the drawback of saving time for later should be more fatal at critical book. So I expect this method to bias results the other way (in favor of borderline books). Still very interesting test, and might even optimise slowmover as side effect. 
2 sensible opposing factors: with bigger NN less thinking time is required early, with critical borderline book more.

**Comment by xoto10** [comment]
My apologies: I'm looking at the code, and actually it doesn't work how I thought, in fact almost the opposite.

The basic calculation doesn't use timeLeft (which is adjusted by the Slow Mover option), just the ply number of the game - time use at the start of the game will be unchanged by Slow Mover under normal tc.

timeLeft is used indirectly to limit how much of the current available time is chosen as optimum on the current move. Reducing timeLeft with Slow Mover = 80 effectively raises the cap from 0.2 * available time to 0.25 * available time, but this only has effect when time is low (basically, playing on increment). Lower values for Slow Mover allow stockfish to take too much time in tricky positions when time is low, making time trouble more likely on later moves.

**Comment by vdbergh** [comment]
| Name | Draw ratio | Bias(Elo) | Elo | Normalized Elo | Test |
|-|-|-|-|-|-|
| noob_3moves.epd (LTC)|0.906|22| 9.85 [9.01, 10.68]|  32.85 [30.14, 35.57] |[Test](   https://tests.stockfishchess.org/tests/view/611fdc804977aa1525c9cba6)|
| bjbraams_chessdb_198350_lines.epd (LTC)|0.69|118| 12.52 [11.30, 13.76]|  28.30 [25.51, 31.08] |[Test](https://tests.stockfishchess.org/tests/view/612095c44977aa1525c9cc07)|
| UHO_XXL_+0.90_+1.19.epd (LTC)|0.49|200| 12.62 [11.24, 13.99]|  25.62 [22.83, 28.41]|[Test](https://tests.stockfishchess.org/tests/view/611fdd2b4977aa1525c9cba9)|

**Comment by vdbergh** [comment]
Above is a test of engines that are closer together in strength. Unfortunately the result challenges some of the earlier assumptions. 

1. It seems results for engines that are far apart in strength cannot be extrapolated to engines that are close together in strength. 
2. The BayesElo (Rao-Kupper) model does not predict correctly the effect of biased openings.

This test should be regarded as providing no evidence for the fact that biased opening books are better. IMHO it means that the book change should be put on hold until more evidence is available (all things being equal one prefers a balanced book I think). 

It is conceivable that the noob book benefits from selection bias, but if so, one would have expected this is to be also visible in the test in the opening post, where it wasn't.

It has been suggested to do a test with time odds. Cutechess-cli supports time odds but it would require some (minor) changes in fishtest to make this available. I can do this but it will take some time, Personally I am quite skeptical about the slowmover approach.

**Comment by vondele** [comment]
thanks for testing.. btw, let's keep time-odds out of the discussion for now. I'm fine to have that in a separate thread.

It is interesting how this test also challenges the nElo results of some other tests (https://github.com/official-stockfish/Stockfish/issues/3323#issuecomment-900857490 and https://github.com/official-stockfish/Stockfish/issues/3323#issuecomment-900628924). 

I would say that from a nElo perspective the books are actually remarkably similar.

Concerning the preference for a balanced book, I'm not sure, it is just a different 'selection bias', we will favor a bit patches that perform better in those 'sharp' situations. We can't predict (and hard to test) what the result of such a different 'evolutionary pressure' will be. However, I think it would be nice to try out the change. I'm sure we'll continue to see progress, it will be just a different direction.

I have a slight preference for the UHO set, the reason being that they have been scored with a different engine, so in some sense introduce a bit more variability.

**Comment by vdbergh** [comment]
Well I think the Elo results from SPRT tests are too unreliable to have any hope of obtaining a significant result (note that that the confidence interval for nElo is larger than for Elo, even in relative terms). So I would regard these results as anecdotic at best. 

I strongly prefer fixed length tests. But if one really wants to test single commits (rather than an aggregate of multiple commits) one needs much more than 60000 games to get a somewhat reliable result (I would go for 200000). And even then there is the risk that this commit is somehow "special". So one would have to repeat the test for several commits. Maybe there is a more sophisticated way of approaching this from a statistical point of view, but it is not clear to me.  

Anyway. It could be a good idea to introduce an unbalanced book and then to redo the test after a few months. It if turns out that the noob book is still significantly better, then we can switch back (since then the result can no longer be attributed to selection bias).

**Comment by vondele** [comment]
yes, I think that's a good plan, I'll make a PR for fishtest with the new book, and we see in a few months what the result was.

**Comment by vondele** [comment]
default book updated to UHO_XXL_+0.90_+1.19.epd 

thanks all for book contributions and testing.

**Comment by NKONSTANTAKIS** [comment]
@vdbergh Since your computations for Davidson model / Bayes elo found 50% draw-rate to be optimal, and indeed it looks to be/makes sense, but the lowered draw-rate brings down normalized elo, I would like to ask if there is a clean (or easily approximative) way to compute optimal for normalised elo.

I also think its a good idea to try this other extreme for some months, rapidly getting rid of previous selection bias along the way, but ultimately imo the golden balance is most likely to lay somewhere in the middle, around 70%. (DD to WL ratio around 2:1)

Anyway this for the future, happy respins and have fun!

**Comment by NKONSTANTAKIS** [comment]
STC's seem to be taking a lot more games to resolve. Could we also have an STC closed strength measurement for normalized elo comparison? If the difference is indeed big, then some adjustments might be needed. For LTC nElo is just 22% less, (minus selection bias) so no worries.

**Comment by Vizvezdenec** [comment]
STCs are not taking a lot more games to resolve.
It's an illusion created by the fact that most running STCs are respins of patches that passed STCs previously so they just don't fail fast as your average fishtest patch.

**Comment by Sopel97** [comment]
Regarding the bias in the books... could this be countered by mirroring every other position (+ flip side to move)?

**Comment by NKONSTANTAKIS** [comment]
STC's definitely take much longer than before, mainly due to lowered nElo spread. It has similar effect to narrowing bound width. The respins visualy exaggerate the effect. Whenever a drastic change happens all previous statistics change. Bounds need to be recalibrated or at least measured. Ihttps://tests.stockfishchess.org/tests/view/612286dccb26afb615a5e769
400K to resolve. Its similar situation with HCE testing once we set bounds for LTC NNUE, often 600K unresolved fauzi tune I recall. This was due to the much lowered drawrate of HCE. Now its the same with STC.

**Comment by NKONSTANTAKIS** [comment]
The testing data is here, there is just one data point missing. Normalized elo: 
Wide elo gap: 
LTC:  97 noob_3moves 118 UHO_0.9
STC: 114.2                     112.3
Narrow elo:
LTC: 32.85                      25.62
STC: ?                               ?

**Comment by vdbergh** [comment]
The STC tests should not take more time since the SPRT bounds are expressed in normalized Elo. The worst case expected duration is still 116K games. 

The right tail of the distribution falls off more slowly than for the normal distribution (the falloff is exponential). It follows that very long tests will happen more frequently than one would expect under the normal distribution. But the average duration should stay below 116K games.

**Comment by vdbergh** [comment]
> Regarding the bias in the books... could this be countered by mirroring every other position (+ flip side to move)?

Do I understand correctly that when applied to the starting position, this procedure would yield the starting position with black to move? I guess it depends on the NN training whether the NN can handle such positions sensibly.

I actually understand why in Stehan Pohl's books the bias is always for white and nearly constant. This allows one to use Ordo or BayesElo. These programs use models with a "WhiteAdvantage" parameter. The models assume that the White Advantage is constant. If this assumption does not hold then the programs give incorrect results.

The pentanomial statististics (for head to head matches) on the other hand do not depend on any model so there is no constraint on the bias for the individual positions.

**Comment by Sopel97** [comment]
> Do I understand correctly that when applied to the starting position, this procedure would yield the starting position with black to move? I guess it depends on the NN training whether the NN can handle such positions sensibly.

The net would see that position exactly the same as it sees the start position. The transformation I described is invariant from the perspective of the net. If anything would change it would be from search; and also, hackily, bias could be brought to 0.

**Comment by vdbergh** [comment]
> > Do I understand correctly that when applied to the starting position, this procedure would yield the starting position with black to move? I guess it depends on the NN training whether the NN can handle such positions sensibly.
> 
> The net would see that position exactly the same as it sees the start position. The transformation I described is invariant from the perspective of the net. If anything would change it would be from search; and also, hackily, bias could be brought to 0.

I guess you mean that the average bias would be zero. I agree with that. Note that the RMSbias would not change (this is the square root of the average of the squared biases). So one would still need pentanomial statistics.

**Comment by NKONSTANTAKIS** [comment]
@vdbergh Thanks for the reassuring response. So could we say that the benefit of book performance on given bounds exclusively affects the accuracy of resolution? If so then it would mean that wider bounds with better performing book would provide equal confidence on less average games, right? I was under the impression that given bounds provide equal confidence upon resolution, with varying productivity. Book performance measured in nElo.

**Comment by vdbergh** [comment]
A 10% time odds test (60+0.6 vs 66+0.66).

| Name | Draw ratio | Bias(Elo) | Elo | Normalized Elo | Test |
|-|-|-|-|-|-|
| UHO_XXL_+0.90_+1.19.epd (LTC)|0.50|212| 10.94 [9.68, 12.21]|  **24.05** [21.26, 26.84] |[Test](https://tests.stockfishchess.org/tests/view/612d3f7dbb4956d8b78eb5f4)|
| noob_3moves.epd (LTC)|0.95|18|  3.39 [2.78, 4.00]| **15.46** [12.72, 18.20] |[Test](   https://tests.stockfishchess.org/tests/view/612d3fddbb4956d8b78eb5f7)|

The nElo ratio is `24.05/15.6=1.56` with (approximate) confidence interval `[1.22,1.89]`.

This result is quite different from https://github.com/official-stockfish/Stockfish/issues/3323#issuecomment-903233120 . It seems likely that the latter test suffered from selection bias.

**Comment by NKONSTANTAKIS** [comment]
Impressively contrasting result, indicating high deviation of optimization across these very different position subsets.
I find valuable to research the deviation of optimization across TC too, as I recon quite possible that the critical nature of positions has weakened the STC-LTC correlation. But not sure how, any ideas?

Great result regarding UHO productivity, as all testing so far points great scaling with TC (resistance to elo compression), indicating low amount of deterministic results with higher depth searches. (discovered forced wins/draws)
This does not necessarily mean a weak STC performance, but its a concern.
An STC time odds comparison would safeguard that UHO starts from positive standpoint.

Regarding LTC confidence, a simple calculation of Elo to nElo analogies at TC odds test gives us an x2 performance! Assessing the combined STC+LTC confidence is another reason for STC time odds test.

**Comment by vondele** [comment]
We have been using the new book now for somewhat more than 2 months. 

I've started 3 tests with the old noob_3moves.epd, the new UHO_XXL_+0.90_+1.19.epd and the RT 8moves_v3.pgn books, on the patches committed to master since Aug 22nd (SPRT tested on the UHO book). 

https://tests.stockfishchess.org/tests/view/61891ae3d7a085ad008ef282
https://tests.stockfishchess.org/tests/view/61891ac2d7a085ad008ef280
https://tests.stockfishchess.org/tests/view/61891b0ad7a085ad008ef286

this test was triggered by the question if progress on the new book and progress on the old book are comparable.

**Comment by vdbergh** [comment]
Thanks. I will put the results in a table tonight. However to be able to compare with https://github.com/official-stockfish/Stockfish/issues/3323#issuecomment-911982470 it might be useful to also do the tests at LTC. That way we can see if it is the UHO book that benefited from selection bias this time.

**Comment by vdbergh** [comment]
A less costly alternative is to redo https://github.com/official-stockfish/Stockfish/issues/3323#issuecomment-911982470 for STC. Maybe I'll just do that. Then we also get some new scaling information with respect to TC.

**Comment by NKONSTANTAKIS** [comment]
The results show around 50% UHO benefit.

https://tests.stockfishchess.org/tests/view/618a35a1d7a085ad008ef38b
https://tests.stockfishchess.org/tests/view/618a3541d7a085ad008ef388
https://tests.stockfishchess.org/tests/view/618a34d7d7a085ad008ef385

**Comment by bjbraams** [comment]
NKONSTANTAKIS or vdbergh, can you clarify "50% UHO extra benefit"? What measure? The measure that makes sense to me is how many games are needed (all else equal) to decide the sign (only the sign) of the elo difference between two instances of SF.

**Comment by NKONSTANTAKIS** [comment]
By comparing the 3 tests on the 3 different books of same reference version elo spreaded via time odds, soas to not be affected by selection bias of UHO specific optimization, vs the 3 tests of dev vs the old reference version.

We see that the 2 balanced books have almost similar performance, while on UHO book that we optimize for we get 23.4 vs 14.5. 

So, taking everything into account its very near 50%. This is after ~3 months of patches at UHO and is expected to increase.

**Comment by NKONSTANTAKIS** [comment]
In other words dev has became as stronger vs the last version before book swap as if that version had 10% more time against itself at STC and balanced books, and 50% more at UHO.

This means that performance at balanced books also benefited by UHO progress, taking 2/3 of the gains.

**Comment by vondele** [comment]
I will close this as an issue. We have quite successfully adopted UHO as the training book. We did find a recent case where there is quite a difference between the Elo measured on the 8moves and the UHO book (https://github.com/official-stockfish/Stockfish/pull/3937) but overall the experience is good. Several major tournaments have adopted UHO-style books as well.

We can reopen the issue at a later point if useful.

**Comment by vdbergh** [comment]
What these experiments have thought us is that book tests based on prior commits are completely useless because of selection bias. The only objective book tests are time odds tests - like this one https://github.com/official-stockfish/Stockfish/issues/3323#issuecomment-91198247.

**Comment by vondele** [comment]
@vdbergh I'll link to your test of the DFRC book https://tests.stockfishchess.org/tests/view/62c551fe50dcbecf5fc0cfe8 let us know what your interpretation of those results are.
